packages/deployments/contracts/contracts/core/connext/libraries/AssetLogic.sol

```diff
@@ -228,7 +252,114 @@ library AssetLogic {
       IStableSwap pool = s.adoptedToLocalPools[_canonicalId];
       SafeERC20.safeApprove(IERC20(_assetIn), address(pool), _amount);
 
-      return (pool.swapExact(_amount, _assetIn, _assetOut), _assetOut);
+      return (pool.swapExact(_amount, _assetIn, _assetOut, 0), _assetOut);
+    }
+  }
+
+  /**
+   * @notice Swaps assetIn t assetOut using the stored stable swap or internal swap pool
+   * @dev Will not swap if the asset passed in is the adopted asset
+   * @param _canonicalId - The canonical token id
+   * @param _assetIn - The address of the from asset
+   * @param _assetOut - The address of the to asset
+   * @param _amountOut - The amount of the _assetOut to swap
+   * @return The amount of assetIn
+   * @return The address of assetOut
+   */
+  function _swapAssetOut(
+    bytes32 _canonicalId,
+    address _assetIn,
+    address _assetOut,
+    uint256 _amountOut
+  ) internal returns (uint256, address) {
+    AppStorage storage s = LibConnextStorage.connextStorage();
+
+    // Swap the asset to the proper local asset
+
+    uint8 tokenIndexIn = getTokenIndexFromStableSwapPool(_canonicalId, _assetIn);
+    uint8 tokenIndexOut = getTokenIndexFromStableSwapPool(_canonicalId, _assetOut);
+    if (stableSwapPoolExist(_canonicalId)) {
+      // if internal swap pool exists
+      return (s.swapStorages[_canonicalId].swapInternalOut(tokenIndexIn, tokenIndexOut, _amountOut, 0), _assetOut);
+    } else {
+      // Otherwise, swap via stable swap pool
+      IStableSwap pool = s.adoptedToLocalPools[_canonicalId];
+      uint256 _amountIn = pool.calculateSwapOut(tokenIndexIn, tokenIndexOut, _amountOut);
```

Comment:
this is essentially the same thing as having *no* maximum amount in!

This `_swapAssetOut` is essentially a `swapExactOut` function. If the slippage and/or ratio is off-balance sufficiently, I could be getting a very bad deal here, and the only cap to my risk is however much funds I have available in the first place (bc obvi I cannot transfer more tokens than I own). This invites front-running/sandwiching to basically extract all of my tokens, so long as I own < the pool size.

Reply:
I know this method is `internal` ofc, but anytime we are doing an exact-out-swap, we must have a route to exert control over the maximum amountIn to some degree

Reply:
right -- so i guess this is kind of weird. there's a few constraints we have to solve around:

- This is the only case where a transfer has to perform a swap on behalf of the router instead of the user
- There are no external parameters that can be supplied to enforce slippage
- Slippage cannot be strongly enforced because this is called via `handle`, which will mark the message as handled even if the handling fails (i.e. reconcile reverts because slippage is enforced)

So there's a couple of options here:

1. Use the amount the bridge minted as the default slippage/min received. If it exceeds that, pay as much as you can and emit the debt event
2. Have a global max slippage enforced on the contracts. If your transfer exceeds the slippage, then it would do the same as above.

Additionally, we can add a slippage param to the manual debt repay function that is included in the `PortalFacet`.